### DISCUSSION TITLE
or-tools Installation in ubuntu:20.04 docker container

### DISCUSSION THREAD

**Discussion by andreaxe** [google/or-tools#3484] [Build and configuration issues] [answered]
Hi,

i come to ask for help to understand a error i'm having with an already compiled distribution of a C++ program given to me that uses or-tools library.

The **error** i was having while trying to run the program in my container is the following:

"/DdVC.out: error while loading shared libraries: libortools.so.9: cannot open shared object file: No such file or directory"

Figured i didn't had or-tools installed so i followed the installation instructions and ended up doing the following steps:

1. cloned the stable repository
2. installed the dependencies
3. execute `make cpp` in the cloned directory

In the end i noticed my `/usr/local/lib` and `/usr/local/include` didn't had any files like mentioned here: 
https://groups.google.com/g/or-tools-discuss/c/j6pcNRzah0w
Tried anyway to run the program and the error persisted.

So what i ended up doing was copying all the content of  the generated folder `install_make` to inside my `/usr` folder.
When i run the file again it shows **another error**:

"/DdVC.out: symbol lookup error: ./DdVC.out: undefined symbol: _ZN4absl12lts_2022062319ToInt64MillisecondsENS0_8DurationE"

**Questions**:
Did the procedure of copying the files of `install_make` to inside my `/usr` folder correctly installed the library? 
Can this _symbol lookup error_ be somehow related to any miss doing with the steps i mentioned above trying to install or-tools? 

The person who gave the compiled C++ distribution to me can run the program in his ubuntu:20.04.

Thanks!

ps: all the steps were done inside my ubuntu:20.04 docker container.

**Comment by Mizux** [comment, marked as answer]
```sh
c++filt _ZN4absl12lts_2022062319ToInt64MillisecondsENS0_8DurationE
absl::lts_20220623::ToInt64Milliseconds(absl::lts_20220623::Duration)
```

Do you have something like this ?
https://github.com/google/or-tools/blob/main/tools/docker/images/ubuntu-20.04.Dockerfile

Please note absl API is C++ dialect dependent, since we build or-tools using `-std=c++17` you should also build your program with the same flag otherwise you may have "two kinds of absl" which could explain this error...
If you can provide the command line used to build `./DdVC.out`

**Reply by andreaxe** [reply]
Thanks for your reply.
I was using my own version of a Dockerfile. 
This last attempt I tried to use your Dockerfile (tools/docker: deep refactor (#3200) SHA: cc3fa1974e6a27334652322a411b4c0ae0629de3) with a few tweaks to integrate my REST application (in attach my Dockerfile as txt).
[Dockerfile.txt](https://github.com/google/or-tools/files/9728810/Dockerfile.txt)

![image](https://user-images.githubusercontent.com/9929973/194426320-3c09ff2b-85f9-4709-898c-c2d1d38c7b3d.png)

The container built correctly (the web server is waiting for connections) but when i went inside it to try running the C++ binary i still get the error (as if the library is not installed):
./DdVC.out: error while loading shared libraries: libortools.so.9: cannot open shared object file: No such file or directory

Again what i resort to do was to copy the contents of the `/root/or-tools/install_make` to `/usr` folder and run again. Same error as before.
./DdVC.out: symbol lookup error: ./DdVC.out: undefined symbol: _ZN4absl12lts_2022062319ToInt64MillisecondsENS0_8DurationE

The binary file "DdVC.out" was built using using Visual Studio, which compiles each source file as follows:

`g++ -c -x c++ amgl.cpp -g1 -o amgl.o -Wall -Wswitch -W"no-deprecated-declarations" -W"empty-body" -Wconversion -W"return-type" -Wparentheses -W"no-format" -Wuninitialized -W"unreachable-code" -W"unused-function" -W"unused-value" -W"unused-variable" -O3 -fno-strict-aliasing -fomit-frame-pointer -DUNIX -D_CRT_SECURE_NO_WARNINGS -D_CHANGE_SYS_TIME -DNDEBUG -fthreadsafe-statics -fexceptions -frtti -std=c++17
`

Then, its links using:

`g++ -o DdVC.out -Wl,--no-undefined -Wl,-z,relro -Wl,-z,now -Wl,-z,noexecstack -pthread -ldl -fopenmp amgl.o amgl2.o sqlite3.o main.o -lboost_serialization -lboost_date_time -lboost_system -lboost_log -lboost_thread -lboost_log_setup -ldl -lortools`

Basically i'm trying to deploy a REST API (FastAPI) responsibile to execute the Binary file on behalf of the client and return a response through the REST API. 

Thanks!

**Reply by Mizux** [reply]
I guess you need to link to all absl static libs too, my point is beside ortools link to absl all symbols not used are dropped but since we have absl in our public headers when using ortools you may need symbols from absl which was dropped...
EDIT: seems libortools.so do contains the symbols `absl::lts_20220623::ToInt64Milliseconds(absl::lts_20220623::Duration)`

```sh
%ls install_make/lib/libabsl*
install_make/lib/libabsl_bad_any_cast_impl.a	  install_make/lib/libabsl_flags.a			     install_make/lib/libabsl_log_severity.a				install_make/lib/libabsl_scoped_set_env.a
install_make/lib/libabsl_bad_optional_access.a	  install_make/lib/libabsl_flags_commandlineflag.a	     install_make/lib/libabsl_low_level_hash.a				install_make/lib/libabsl_spinlock_wait.a
install_make/lib/libabsl_bad_variant_access.a	  install_make/lib/libabsl_flags_commandlineflag_internal.a  install_make/lib/libabsl_malloc_internal.a			install_make/lib/libabsl_stacktrace.a
install_make/lib/libabsl_base.a			  install_make/lib/libabsl_flags_config.a		     install_make/lib/libabsl_periodic_sampler.a			install_make/lib/libabsl_status.a
install_make/lib/libabsl_city.a			  install_make/lib/libabsl_flags_internal.a		     install_make/lib/libabsl_random_distributions.a			install_make/lib/libabsl_statusor.a
install_make/lib/libabsl_civil_time.a		  install_make/lib/libabsl_flags_marshalling.a		     install_make/lib/libabsl_random_internal_distribution_test_util.a  install_make/lib/libabsl_strerror.a
install_make/lib/libabsl_cord.a			  install_make/lib/libabsl_flags_parse.a		     install_make/lib/libabsl_random_internal_platform.a		install_make/lib/libabsl_str_format_internal.a
install_make/lib/libabsl_cord_internal.a	  install_make/lib/libabsl_flags_private_handle_accessor.a   install_make/lib/libabsl_random_internal_pool_urbg.a		install_make/lib/libabsl_strings.a
install_make/lib/libabsl_cordz_functions.a	  install_make/lib/libabsl_flags_program_name.a	     install_make/lib/libabsl_random_internal_randen.a			install_make/lib/libabsl_strings_internal.a
install_make/lib/libabsl_cordz_handle.a		  install_make/lib/libabsl_flags_reflection.a		     install_make/lib/libabsl_random_internal_randen_hwaes.a		install_make/lib/libabsl_symbolize.a
install_make/lib/libabsl_cordz_info.a		  install_make/lib/libabsl_flags_usage.a		     install_make/lib/libabsl_random_internal_randen_hwaes_impl.a	install_make/lib/libabsl_synchronization.a
install_make/lib/libabsl_cordz_sample_token.a	  install_make/lib/libabsl_flags_usage_internal.a	     install_make/lib/libabsl_random_internal_randen_slow.a		install_make/lib/libabsl_throw_delegate.a
install_make/lib/libabsl_debugging_internal.a	  install_make/lib/libabsl_graphcycles_internal.a	     install_make/lib/libabsl_random_internal_seed_material.a		install_make/lib/libabsl_time.a
install_make/lib/libabsl_demangle_internal.a	  install_make/lib/libabsl_hash.a			     install_make/lib/libabsl_random_seed_gen_exception.a		install_make/lib/libabsl_time_zone.a
install_make/lib/libabsl_examine_stack.a	  install_make/lib/libabsl_hashtablez_sampler.a	     install_make/lib/libabsl_random_seed_sequences.a
install_make/lib/libabsl_exponential_biased.a	  install_make/lib/libabsl_int128.a			     install_make/lib/libabsl_raw_hash_set.a
install_make/lib/libabsl_failure_signal_handler.a  install_make/lib/libabsl_leak_check.a		     install_make/lib/libabsl_raw_logging_internal.a
```
This `install_make/lib/libabsl_time.a` seems to be a good candidate to start with.

Concerning the `install_make/` vs `/usr` If I were you I would use the CMake build system directly and **without** any  [`CMAKE_INSTALL_PREFIX`](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html) thus it should automatically install it in /usr.
Also for your program I would use a CMake based build so you can let the build system generator adding the correct list of link flags and dependencies (an otherway will be to compile an example in a cpp prebuilt artifact and look at the command line)

**Comment by Mizux** [comment]
Small test trying to run the tsp examples from the cpp Ubuntu-20.04 archive
Dockerfile:
```dockerfile
# ref: https://hub.docker.com/_/ubuntu
FROM ubuntu:20.04 AS env

ENV DEBIAN_FRONTEND=noninteractive
RUN apt-get update \
&& apt-get install -yq wget build-essential cmake zlib1g-dev \
&& apt-get clean \
&& rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
ENTRYPOINT ["/bin/bash", "-c"]
CMD ["/bin/bash"]

# Install CMake 3.24.2
RUN ARCH=$(uname -m) \
&& wget -q "https://cmake.org/files/v3.24/cmake-3.24.2-linux-${ARCH}.sh" \
&& chmod a+x cmake-3.24.2-linux-${ARCH}.sh \
&& ./cmake-3.24.2-linux-${ARCH}.sh --prefix=/usr/local/ --skip-license \
&& rm cmake-3.24.2-linux-${ARCH}.sh

FROM env AS devel
WORKDIR /root
RUN wget https://github.com/google/or-tools/releases/download/v9.4/or-tools_amd64_ubuntu-20.04_cpp_v9.4.1874.tar.gz
RUN tar xzvf or-tools_amd64_ubuntu-20.04_cpp_v*.tar.gz \
&& rm -rf or-tools_amd64_ubuntu-20.04_cpp_v*.tar.gz

FROM devel as build
RUN cd or-tools_cpp_Ubuntu-20.04-* \
&& cd examples/tsp \
&& cmake -S. -Bbuild -DCMAKE_PREFIX_PATH=/root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874 \
&& cmake --build build -v \
&& ./build/bin/tsp
```

output:
```sh
docker build --tag=plop --target=build .
...
[ 50%] Building CXX object CMakeFiles/tsp.dir/tsp.cc.o
/usr/bin/c++ -DHAVE_CONFIG_H -DOR_TOOLS_AS_DYNAMIC_LIB -DUSE_BOP -DUSE_CBC -DUSE_CLP -DUSE_GLOP -DUSE_LP_PARSER -DUSE_PDLP -DUSE_SCIP -I/root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/examples/tsp -isystem /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/include -isystem /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/include/coin -isystem /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/include/eigen3 -O3 -DNDEBUG -fPIE -fwrapv -std=c++17 -MD -MT CMakeFiles/tsp.dir/tsp.cc.o -MF CMakeFiles/tsp.dir/tsp.cc.o.d -o CMakeFiles/tsp.dir/tsp.cc.o -c /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/examples/tsp/tsp.cc
[100%] Linking CXX executable bin/tsp
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/tsp.dir/link.txt --verbose=1
/usr/bin/c++ -O3 -DNDEBUG CMakeFiles/tsp.dir/tsp.cc.o -o bin/tsp  -Wl,-rpath,/root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libortools.so.9.4.1874 -ldl /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_parse.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_usage.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_usage_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_marshalling.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_reflection.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_config.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_private_handle_accessor.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_commandlineflag.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_commandlineflag_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_flags_program_name.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_distributions.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_seed_sequences.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_internal_pool_urbg.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_internal_randen.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_internal_randen_hwaes.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_internal_randen_hwaes_impl.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_internal_randen_slow.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_internal_platform.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_internal_seed_material.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_random_seed_gen_exception.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_raw_hash_set.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_hashtablez_sampler.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_hash.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_city.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_low_level_hash.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_leak_check.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_statusor.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_status.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_cord.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_cordz_info.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_cord_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_cordz_functions.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_exponential_biased.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_cordz_handle.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_bad_optional_access.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_bad_variant_access.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_str_format_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_synchronization.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_stacktrace.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_symbolize.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_debugging_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_demangle_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_graphcycles_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_malloc_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_time.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_strings.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_strings_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_base.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_spinlock_wait.a -lrt /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_throw_delegate.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_int128.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_civil_time.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_time_zone.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_bad_any_cast_impl.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_raw_logging_internal.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libabsl_log_severity.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libprotobuf.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libre2.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libCbcSolver.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libOsiCbc.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libCbc.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libCgl.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libClpSolver.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libOsiClp.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libClp.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libOsi.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libCoinUtils.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libscip.a /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/lib/libz.a -pthread -lpthread 
make[2]: Leaving directory '/root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/examples/tsp/build'
[100%] Built target tsp
make[1]: Leaving directory '/root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/examples/tsp/build'
/usr/local/bin/cmake -E cmake_progress_start /root/or-tools_cpp_Ubuntu-20.04-64bit_v9.4.1874/examples/tsp/build/CMakeFiles 0
WARNING: Logging before InitGoogleLogging() is written to STDERR
I1007 07:11:22.074460   113 tsp.cc:74] Objective: 4384
I1007 07:11:22.074573   113 tsp.cc:77] Route for Vehicle 0:
I1007 07:11:22.074586   113 tsp.cc:86] 0 -> 9 -> 5 -> 8 -> 6 -> 2 -> 10 -> 16 -> 14 -> 13 -> 12 -> 11 -> 15 -> 3 -> 4 -> 1 -> 7 -> 0
I1007 07:11:22.074590   113 tsp.cc:87] Distance of the route: 4384m
I1007 07:11:22.074593   113 tsp.cc:88] 
I1007 07:11:22.074597   113 tsp.cc:89] Advanced usage:
I1007 07:11:22.074600   113 tsp.cc:90] Problem solved in 2ms
Removing intermediate container 25544f88582d
 ---> 872c8d5ca089
Successfully built 872c8d5ca089
Successfully tagged plop:latest
```
Please take a look at the `Linking CXX executable bin/tsp` step and try to reproduce the same link command...

**Comment by Mizux** [comment]
It seems we didn't link against absl_time...
EDIT: Just add it on `main` branch see https://github.com/google/or-tools/commit/a846062054d3c9f68be7a47e6fafdb805751e1a5

https://github.com/google/or-tools/blob/69ab636173b9078710eda628397eddd85d1ba53e/cmake/deps.cmake#L35-L56

https://github.com/google/or-tools/blob/69ab636173b9078710eda628397eddd85d1ba53e/cmake/cpp.cmake#L154-L167

@andreaxe could you try to add `absl::time` and retest ?

**Reply by Mizux** [reply]
devNote:
main:
```sh
$ git grep -n absl::ToInt64Milliseconds
ortools/constraint_solver/constraint_solver.cc:1552:  return absl::ToInt64Milliseconds(timer_->GetDuration());
ortools/constraint_solver/constraint_solver.h:4421:               : absl::ToInt64Milliseconds(duration_limit());
ortools/linear_solver/linear_solver.h:833:               : absl::ToInt64Milliseconds(time_limit_);
ortools/linear_solver/linear_solver.h:846:    return absl::ToInt64Milliseconds(DurationSinceConstruction());
ortools/math_opt/solvers/glpk_solver.cc:419:    const int64_t time_limit_ms = absl::ToInt64Milliseconds(

$ nm install_make/lib/libortools.so | grepc "absl.*ToInt64Milliseconds.*Duration"
000000000163b6a0 T _ZN4absl12lts_2022062319ToInt64MillisecondsENS0_8DurationE
```

main + `absl::time` in `cmake/deps.cmake`
```sh
$ nm install_make/lib/libortools.so | grepc "absl.*ToInt64Milliseconds.*Duration"
0000000000e7faa0 T _ZN4absl12lts_2022062319ToInt64MillisecondsENS0_8DurationE
$ nm -C install_make/lib/libortools.so | grepc "absl.*ToInt64Milliseconds.*Duration"
0000000000e7faa0 T absl::lts_20220623::ToInt64Milliseconds(absl::lts_20220623::Duration)
```
